File: 0322-coin-change/solution.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @cache
        def coin(left):
            if left == 0:
                return 0
            if left <= 0:
                return float('inf')
            lowest = float('inf')
            for cur in coins:
                res = coin(left - cur)
                lowest = min(lowest, res + 1)
            return lowest
        res = coin(amount)
        if res == float('inf'):
            return -1
        return res

File: 0322-coin-change/solution.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, amount + 1);
        dp[0] = 0;
        for(int i = 1; i < amount + 1; ++i){
            for(auto cur : coins){
                if(i >= cur){
                    dp[i] = min(dp[i], dp[i - cur] + 1);
                }
            }
        }
        if(dp[amount] == amount +1){
            return -1;
        }
        return dp[amount];
    }
};
